arch/arc/core/thread_entry_wrapper.S

```diff
@@ -35,3 +35,24 @@ z_thread_entry_wrapper1:
 	POPR r0
 	j z_thread_entry
 	nop
+
+
+#if !defined(CONFIG_MULTITHREADING)
+GTEXT(z_main_no_multithreading_entry_wrapper)
+
+/* void z_main_no_multithreading_entry_wrapper(*p1, *p2, *p3, *main_stack, *main_entry) */
+SECTION_FUNC(TEXT, z_main_no_multithreading_entry_wrapper)
+	MOVR sp, r3
+
+	/* *p1, *p2, *p3 are in r0, r1, r2 already */
+	jl [r4]
+	nop
+
+	/*
+	 * If we return from main we'll return from main wrapper and appear here.
+	 * Go to infinite loop as there is nothing more to do.
+	 */
+z_main_no_multithreading_entry_wrapper_end:
+	b z_main_no_multithreading_entry_wrapper_end
```

Comment:
Yep, this may look a bit strange for real use cases. This was added for Zephyr tests - as they do leave main() even in case of multithreading is disabled. And (as I understand) it was done intentionally - we need to leave main as we do reporting of coverage info, etc. after that.

The overhead here is one 2-byte instruction - it's anyway better than just allow to execute some random code if we return from main.